### ISSUE TITLE
上位解法

### ISSUE THREAD

**Issue by riow1983** [riow1983/nbme-score-clinical-patient-notes#3]
上位解法のキーテクニックは次の3つ:
- MLM
- PL学習 (リーク防止PL学習, 多重PL学習, MPL学習)
- Adversarial Training (AWP, etc.)

![key techniques](https://github.com/riow1983/nbme-score-clinical-patient-notes/blob/main/png/key_techniques.png)

**Comment by riow1983** [comment]
[MLM]
- Paper: 
  - [Don’t Stop Pretraining: Adapt Language Models to Domains and Tasks](https://arxiv.org/pdf/2004.10964.pdf)
- Code:
  - [[Coleridge] BERT - Masked Dataset Modeling](https://www.kaggle.com/code/tungmphung/coleridge-bert-masked-dataset-modeling)
  - [NBME MLM](https://www.kaggle.com/code/currypurin/nbme-mlm/notebook)

未ラベルデータ (patient_notes.csv)でMLM訓練を実施し, 訓練後モデルを保存している.
モデルは`model = AutoModelForMaskedLM.from_pretrained(model_name, config=config)`で読みこみ, 
```python
data_collator = DataCollatorForLanguageModeling(tokenizer=tokenizer, mlm_probability=0.15)
trainer = Trainer(
        model=model,
        args=training_args,
        train_dataset=tokenized_datasets["train"],
        eval_dataset=tokenized_datasets['valid'],
        data_collator=data_collator,
    )
trainer.train()
```
としてmaskingと訓練を実行している. あくまで標準的なMLM訓練なので全て:hugs: Trainerで完結する.

**Comment by riow1983** [comment]
[PL学習]
- リーク防止学習: 
  - Discussion: https://www.kaggle.com/c/google-quest-challenge/discussion/129840
  - Diary: https://github.com/riow1983/nbme-score-clinical-patient-notes#2022-04-10
- 多重PL学習: teacherの学習epoch → teacherによるPLによるstudentの学習epoch → studentによるPLによるstudentの再学習epoch → studentによるPLによるstudentの再再学習epoch → ...という方式
>3重PL学習の例:
1st model - only train labels
2nd model - train labels + pseudo labels of 1st model
3rd model - train labels + pseudo labels of 2nd model
4th model - train labels + pseudo labels of 3rd model <- This 
https://www.kaggle.com/competitions/nbme-score-clinical-patient-notes/discussion/322799#1776423
- MPL学習:
  - Paper:
    - [Meta Pseudo Labels](https://arxiv.org/pdf/2003.10580.pdf)
  - Code:
    - [playground for meta-pseudo-label](https://www.kaggle.com/code/hengck23/playground-for-meta-pseudo-label)

この内MPLが最も効果があったと3rdチームが証言していることから, 以下はMPLについて:
MPL擬似アルゴリズム: https://github.com/riow1983/nbme-score-clinical-patient-notes/blob/main/png/MPL.png
以下3rdチームが参考にした実装について
イテレーション内(通常はdataloaderのループ配下)で, 
1/4 teacherが生成したPL...
2/4 ...PLで訓練したstudentの...
3/4 ...studentの変化幅...
4/4 ...変化幅が係数のようについたPL BCE lossでteacherが訓練されている.
これはイテレーション on the flyでPLが生成され, そのPLがstudentの学習に即座に供与されているため, teacherの学習epoch done → studentの学習epoch done → teacherの再学習epoch done → studentの再学習eooch doneという方式よりも処理オーバーヘッドが少なくて良い.
```Python
# experiment3: meta pseudo label
# https://github.com/kekmodel/MPL-pytorch/blob/main/main.py

l_x = torch.from_numpy(x_train).float().cuda()
l_y = torch.from_numpy(y_train).float().cuda()


teacher = Net().cuda()
student = Net().cuda()
t_optimizer = optim.SGD(teacher.parameters(),lr=0.0005, momentum=0.9)
s_optimizer = optim.SGD(student.parameters(),lr=0.0005, momentum=0.9)

for iteration in range(200):
    # subscript: t,s : teacher,student
    # subscript: l,u : label,unlabel
    
    teacher.train()
    student.train()
    t_optimizer.zero_grad()
    s_optimizer.zero_grad()
    
    random_sample = np.random.choice(num_unlabel*2,16).tolist()
    u_x = torch.from_numpy(x_unlabel[random_sample]).float().cuda()
 
    
    #prepare other input
    #note these is detached, i.e. not used for backprop
    s_l_logit = student(l_x)
    s_l_loss = F.binary_cross_entropy_with_logits(s_l_logit.detach(), l_y)
    
    # 1/4 teacherが生成したPL...
    t_u_logit = teacher(u_x)
    pseudo_y =  (t_u_logit.detach()>0).float()
    #------
    
    
    
    # 2/4 ...PLで訓練したstudent...
    #train student : update student using pesudo label data only
    s_u_logit = student(u_x)
    s_u_loss  = F.binary_cross_entropy_with_logits(s_u_logit, pseudo_y)
    s_u_loss.backward()
    s_optimizer.step()
    
    # 3/4 ...studentの変化幅...
    #train teacher : update teacher using pesudo label data student change in loss
    s_l_logit_new = student(l_x)
    s_l_loss_new  = F.binary_cross_entropy_with_logits(s_l_logit_new.detach(), l_y)
    change = s_l_loss_new - s_l_loss
    #https://github.com/kekmodel/MPL-pytorch/issues/6
    
    t_l_logit = teacher(l_x)
    t_l_loss = F.binary_cross_entropy_with_logits(t_l_logit, l_y)
    # 4/4 ...変化幅が係数のようについたPL BCE lossでteacherが訓練されている.
    t_mpl_loss  = change * F.binary_cross_entropy_with_logits(t_u_logit, pseudo_y)
    (t_l_loss+t_mpl_loss).backward()
    t_optimizer.step()
```

**Comment by riow1983** [comment]
[Adversarial Training]
- AWP
- etc.

AWP #4 および社内論文発表会資料参照.
